[WIP] [AMD] Linear layout conversion for PaddedSharedEncodingAttr

alefimov-amd opened the pull request:
This PR adds basic conversion for padded encoding in order to enable memdesc_* operations to use it.

lezcano reviewed (changes requested):
The issue is that the padded layout is not a linear layout so I think it's better to add ad-hoc paths whenever this sort of layouts are used, as otherwise some generic LL code could be silently wrong if you pass a PaddedLayout

antiagainst commented:
Right. I let Alek know earlier internally this is not the proper way to go. He will rework it.

alefimov-amd commented:
@lezcano Hi!

You are right, it is impossible to express paddings in linear layout, but there are a lot of infrastructure similar for both padded and swizzled layouts, so I still want to try reuse it as much as possible 🙂 

I've implemented two approaches at the moment:

## This PR

I've modified generation of linear layout for padded encoding slightly, so instead of mapping *"offset"* -> "dim1", "dim2", it maps *"index"* -> "dim1", "dim2". *"offset"* precisely defines relative offset between elements, *"index"* defines only order of elements in memory, allowing later stages to insert arbitrary paddings between elements.

This approach gives a mechanism to distinguish padded and swizzled layouts after conversion to LL, so additional checks could be implemented and if algorithm already uses dimension names it will detect possible issue.

## PR #7944

It implements ad-hoc paths, like we did before, so If you think previous idea is not safe enough, let's go with it.

lezcano commented:
We can reuse this infrastructure, but as pointed out in https://github.com/triton-lang/triton/pull/7944, one has to be careful. The way to reuse it is by treating this as a different beast than LinearLayouts (because it is!) and reuse what can be reused from the lowerings but being careful when doing so. This has to be audited case by case, not by simply creating an associated LinearLayout to the padded layout and using it in there.

In fact, whenever we call in a lowering `toLinearLayout` and we use it you will have a place where you'll have to figure out how does padding affect to that logic, as you will virtually alwayas need to write special logic for padding layouts around that code.

ThomasRaoux commented:
> We can reuse this infrastructure, but as pointed out in #7944, one has to be careful. The way to reuse it is by treating this as a different beast than LinearLayouts (because it is!) and reuse what can be reused from the lowerings but being careful when doing so. This has to be audited case by case, not by simply creating an associated LinearLayout to the padded layout and using it in there.
> 
> In fact, whenever we call in a lowering `toLinearLayout` and we use it you will have a place where you'll have to figure out how does padding affect to that logic, as you will virtually alwayas need to write special logic for padding layouts around that code.

Fully agree with @lezcano here. Actually I don't think we should be allowed to call `toLinearLayout` on a padded layout.
